services/src/main/java/org/keycloak/services/validation/Validation.java

```diff
@@ -87,9 +87,13 @@ private static void addError(List<FormMessage> errors, String field, String mess
     }
 
     public static List<FormMessage> validateUpdateProfileForm(RealmModel realm, MultivaluedMap<String, String> formData) {
+        return validateUpdateProfileForm(realm, formData, realm.isEditUsernameAllowed());
+    }
+    
+    public static List<FormMessage> validateUpdateProfileForm(RealmModel realm, MultivaluedMap<String, String> formData, boolean userNameRequired) {
         List<FormMessage> errors = new ArrayList<>();
         
-        if (!realm.isRegistrationEmailAsUsername() && realm.isEditUsernameAllowed() && isBlank(formData.getFirst(FIELD_USERNAME))) {
+        if (!realm.isRegistrationEmailAsUsername() && userNameRequired && isBlank(formData.getFirst(FIELD_USERNAME))) {
```

Comment:
I found this rather confusing and it actually took me quite a bit of time to decipher what's going on here. 

I would change `userNameRequired` to `validateUsername`. Then change `!realm.isRegistrationEmailAsUsername() && userNameRequired` to `validateUsername`, and add `!realm.isRegistrationEmailAsUsername()` to L90. That'd make it a lot more obvious what's going on IMO.

Reply:
Sure we can change. But It is just parametrizing whether or not the username is required instead of always obtaining it from the realm. Is that complicated ?